### PR TITLE
week2 20기 분석 정재원

### PR DESCRIPTION
**Pull request by havehill**
수정하여 다시 올립니다!

### CODE REVIEW

**Comment by okaybody10** [comment]
knn.ipynb 파일을 보시면 중간에 cs231n/classifiers/k_nearest_neighbor.py 파일에서 함수를 불러오는 부분이 있어욥
ipynb에서는 아마 여기에서 함수를 불러오려고 했을 텐데, 재원님께서는 직접 클래스로 구현을 하셔서 ipynb에서 못받아들이고 있는 것 같아요 ㅠㅠ
클래스를 새로 짠 것도 진짜 대단하신거 같은데,,
뭐 아무튼.. 그 중간에 짜신 함수를 py로 옮기면 아마 코드가 돌아갈거에요
시간 되실 때 한번 수정해보세요 ㅎㅎ

이번 한 주도 고생 많으셨습니다!

**Comment by havehill** [comment]
> knn.ipynb 파일을 보시면 중간에 cs231n/classifiers/k_nearest_neighbor.py 파일에서 함수를 불러오는 부분이 있어욥 ipynb에서는 아마 여기에서 함수를 불러오려고 했을 텐데, 재원님께서는 직접 클래스로 구현을 하셔서 ipynb에서 못받아들이고 있는 것 같아요 ㅠㅠ 클래스를 새로 짠 것도 진짜 대단하신거 같은데,, 뭐 아무튼.. 그 중간에 짜신 함수를 py로 옮기면 아마 코드가 돌아갈거에요 시간 되실 때 한번 수정해보세요 ㅎㅎ
> 
> 이번 한 주도 고생 많으셨습니다!

헉...............그렇군요 제가 문제 이해를 완전 잘못했나봐요ㅠㅠ 스터디 강의 듣고 꼭 다시 한 번 수정해봐야겠어요..! 재혁님도 늦게까지 수고하십니다 ㅠㅠ 감사합니다!!

**Comment by okaybody10** [comment]
안녕하세요! 일단 고생 많으셨습니다 ㅎㅎ
softmax.py 에서 주석화 하신 부분 봤는데 음.. 지금 i행 - i번째 인덱스가 1대1로 연결 되어야 하니까
인덱스만을 가지고 있는 ndarray 하나랑 정답 들고 있는 ndarray를 동시에 indexing에 넣어주면 돼요
말로 하면 어려우니까 답을 적어드리면
```correct = soft_fun[np.arange(N), y]```로 적을 수 있어요.
이거 반영하시면 마무리 될 듯 하네요 ㅎㅎ

File: softmax.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+        return loss_sum + reg_loss
+    
+    loss = softmax(X,W)
+
+
+    # Computing the gradient **NUMERICALLY**
+    grad = np.zeros(W.shape)
+    Lw = softmax(X,W)
+    h = 1e-5
+    it = np.nditer(np.array(W), flags = ['multi_index'], op_flags = ['readwrite'])
```
Comment (okaybody10) [inline softmax.py:60]: 와 이건 뭔가요...? 처음 보네요 nditer..? 신기하네요

File: softmax.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    # Computing the softmax loss - 실패!!!
+    def softmax(x,w):
+        f = x.dot(w)
+        f -= np.max(f)
+
+        soft_fun = np.exp(f) / np.sum(np.exp(f), axis=1)[:, np.newaxis]     
+        correct = np.zeros(y.shape)
+        #for i in range(x.shape[0]):
+        #    correct[i] = soft_fun[i, y[i]]
+        ## 이 for loop을 어떻게 vectorizied해야할지 모르겠습니다하...;;
```
Comment (okaybody10) [inline softmax.py:105]: 지금 i행 - i번째 인덱스가 1대1로 연결 되어야 하니까
인덱스만을 가지고 있는 ndarray 하나랑 정답 들고 있는 ndarray를 동시에 indexing에 넣어주면 돼요
코드는 ```correct = soft_fun[np.arange(N), y]```로 적을 수 있어요.

File: softmax.py [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+    # Store the loss in loss and the gradient in dW. If you are not careful     #
+    # here, it is easy to run into numeric instability. Don't forget the        #
+    # regularization!                                                           #
+    #############################################################################
+    # *****START OF YOUR CODE (DO NOT DELETE/MODIFY THIS LINE)*****
+
+    # Computing the softmax loss - 실패!!!
+    def softmax(x,w):
+        f = x.dot(w)
+        f -= np.max(f)
```
Comment (okaybody10) [inline softmax.py:99]: 동일하게, 굳이 전체 범위에서 최대를 뺄 필요 없고, 행에 대해서 최대를 빼주는게 젤 좋습니다. (범위 줄이기, 전체범위로 최대를 하면 조금 손해를 보는 경우가 있더라구요)
즉, ```np.amax(f, axis = 1, keepdims=True)``` 또는 ```np.max(f, axis=1).reshape(-1, 1)``` 이러허게 구현하시면 좋을 것 같아욥

**Review by okaybody10 — COMMENTED** [review]
Code Review I End
